### ISSUE TITLE
Kopia on docker: privileged, cap_add, security_opt, /dev/fuse required?

### ISSUE THREAD

**Issue by woiwoiwoiwoi** [kopia/kopia#3133]
I'm running Kopia via docker on my QNAP NAS, having set it up following the [docker-compose from the official documentation](https://kopia.io/docs/installation/#docker-images). After noticing it could not read all files and could not mount snapshots, I finally came across the hint in #2534 that I should run it as `privileged: true`. This indeed solved my issues. However, I always thought I should not run containers as privileged from a security perspective.

After searching some more, I found [another official docker-compose here in this repo](https://github.com/kopia/kopia/blob/master/tools/docker/docker-compose.yml) that adds the privileged flag and then three more:

```
      privileged: true
      cap_add:
        - SYS_ADMIN
      security_opt:
        - apparmor:unconfined
      devices:
        - /dev/fuse:/dev/fuse:rwm
```

Does Kopia require all these? If so, why are there two different official docker-compose guidelines - one of which (the more prominent one) does not include them?

And what about the recommendation not to run containers as privileged?

**Comment by jkowalski** [comment]
Kopia runs just fine as a regular user, so if all you want to snapshot are regular files and you don't need to use FUSE mount you should not need to use privileged containers at all and should be able to drop unneeded capabilities. The recommendation would be to try Kopia without any extra privileges, see if anything breaks for your use case and increase privileges only as needed.

**Comment by woiwoiwoiwoi** [comment]
Thank you for that clarification!

Turns out then, that I went about it the right way. As mentioned, I backed up as a regular user for a while until I noticed two problems:
1. some files weren't being backed up.
2. Kopia could not mount snapshots to the filesystem.

Which privileges would be required for these things? For the moment, the container runs as privileged. But if possible, I would like to reduce that to fewer privileges.

What is a FUSE mount?

**Comment by woiwoiwoiwoi** [comment]
I just realised that Kopia is not following symlinks while backing up. I'm very sure that this used to work. To try and fix this I tried all the mentioned docker options, but nothing is working:

```
privileged: true
cap_add:
- SYS_ADMIN
security_opt:
- apparmor:unconfined
devices:
- /dev/fuse:/dev/fuse:rwm
```

Is there a way to have a docker install of Kopia follow symlinks?

**Comment by jkowalski** [comment]
Nope, this was never working, PRs to improve that are always welcome.

**Comment by woiwoiwoiwoi** [comment]
> Nope, this was never working, PRs to improve that are always welcome.

What about hardlinks? Do they work in a docker setup?

**Comment by jlssmt** [comment]
same problem with mounting here.
with default docker-compose from installation docu: https://kopia.io/docs/installation/#docker-images
I expect, that mounting a snapshot is possible due to:

```yaml
        volumes:
            # Mount path for browsing mounted snaphots
            - /path/to/tmp/dir:/tmp:shared
```

But you get `/usr/bin/fusermount3: fuse device not found, try 'modprobe fuse' first` if you press the "mount" button in gui.
Is it really required to forward fuse from host to the container?

**Comment by jeacott1** [comment]
+1 on the docker image including  fuse https://github.com/AppImage/AppImageKit/wiki/FUSE

**Comment by MrModest** [comment]
I believe this first docker-compose file comes from this SO answer: https://stackoverflow.com/a/49021109/7422280

btw, for me these 3 options together made mounting work withour setting `priveleged: true`:

```yaml
      cap_add:
        - SYS_ADMIN
      security_opt:
        - apparmor:unconfined
      devices:
        - /dev/fuse:/dev/fuse:rwm
```

**Comment by davidgeiger** [comment]
Facing the same issue on my Unraid installation. I need to pass the fuse device and set the SYS_ADMIN capability in order to mount snapshots. Otherwise Kopia just creates an empty folder per snapshot and logs: `DEBUG kopia/server mount controller for (...) not found, starting`. Note that I do not need to set the privileged flag or disable apparmor.

@jkowalski: The link provided by @jeacott1 looks promising. Is this a viable alternative to increasing privileges?

> You'll often hear "oh, just add these arguments to docker run --cap-add SYS_ADMIN --cap-add MKNOD --device /dev/fuse:mrw and it'll work", but that is considered to be insecure.
> 
> Instead, just use the --appimage-extract-and-run parameter to the AppImage in your build script:
